### ISSUE TITLE
NotNullIfNotNull not enforced and not inherited in overriding methods

### ISSUE THREAD

**Issue by roji** [dotnet/roslyn#49428]
The compiler doesn't seem to generate warnings if the method implementation violates NotNullIfNotNull (as it does for, say, NotNullWhen).

In addition, when NotNullIfNotNull is defined on a base method, overriding methods do not inherit it, causing issues for polymorphism.

```c#
var inherited = new Inherited();
var asBase = (Base)inherited;

Console.WriteLine(asBase.Foo("hello").Length); // no warning
Console.WriteLine(inherited.Foo("hello").Length); // warning

class Base
{
    [return: NotNullIfNotNull("s")]
    public virtual string? Foo(string? s) => null; // no warning
}

class Inherited : Base
{
    // No warning on nullability of return type not matching overridden member
    public override string? Foo(string? s) => null;
}
```

dotnet SDK: 5.0.100

**Comment by RikkiGibson** [comment]
We implemented a "best-effort" analysis for this attribute which simply warns at the return if the flow state of the input referenced by `[NotNullIfNotNull]` is not-null, but the output is maybe-null.

Regarding the first case: https://github.com/dotnet/roslyn/issues/44539#issuecomment-683968032
It might be possible to give a warning if the entire body is just "return null;" or equivalent, but we probably can't give the right warnings on more sophisticated cases.

I think we can fix override checks though. That would be a follow-up to https://github.com/dotnet/roslyn/issues/44735#issuecomment-675768560

**Comment by jaredpar** [comment]
@jcouv FYI

I believe presently we do not attempt to enforce these attributes in OHI scenarios. The only ones that we track are the `[AllowNull]` and `[NotNull]` ones. I think it would be more of a design change to do this.

**Comment by RikkiGibson** [comment]
We do not currently "enforce" in the sense of giving additional warnings, but there are scenarios where we *remove* warnings.

```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class Base {
    public virtual string M(string x) => x;
}

class Override : Base {
    [return: NotNullIfNotNull("x")] // no warning, but if we removed this attribute, we'd get a warning
    public override string? M(string? x) => x;
}
```

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AxAHYCuANuRMOXAARzHW0CwAUAAIAMdHAjADoAIpggBzYgHtYmAMZQBAYUkATOAEEm5XFExQA3O3YcATHQBC0egG92de7wDMdNJgQxSEcrz48AsgAU/DwAHgCUdAC8AHx0IYZsAL5GnGYA8hgICJhqdCAWVnS2bA50ANocAOz5AHKSMDUU5ACS2HUNTQEARCFdYQC6dAD0Q3RSdADuEAjEmMRiSHTApDB0OJP0CHBgkhgqdDAAFnp0EDAw2cvwixNwAOT7YnCrEJPTs/N2DhzOu4jZuWCAH46IFgXEIjE4glEkA==

**Comment by jaredpar** [comment]
@RikkiGibson I was meaning more do we consider these as a part of determing if an override / implementation is correct? Essentially do we enforce that if the base does / does not have the attribute that the override does or does not. I believe the intent is to only do this for `[AllowNull]` and `[NotNull]`. 

As for whether it should impact the behavior at the call sites I agree with your assesment.

**Comment by RikkiGibson** [comment]
No, we don't require adding a NotNullIfNotNull attribute to an override simply because it is present on the base.

I also do not think that the attribute should be "inherited", but rather, we should enforce as best we can that an override will honor the base's contract. For example:

```cs
class Base
{
    [return: NotNullIfNotNull("s")]
    public virtual string? M(string? s) => s;
}

class Override1 : Base
{
    // warning: return type doesn't meet the requirements of the base method due to a nullability difference
    public override string? M(string? s) => s;
}

class Override2 : Base
{
    // no warning
    [NotNullIfNotNull("s")]
    public override string? M(string? s) => s;
}

class Override3 : Base
{
    // no warning
    public override string M(string? s) => s ?? "a";
}
```

**Comment by jaredpar** [comment]
That feels more like a featuer request though correct? The "enforce the base contract" was a fairly deliberate decision that we made for `MemberNotNull` and that ended  up having a lot of design decisions around it. Doesn't feel like a bug fix level to add this behavior here.

**Comment by RikkiGibson** [comment]
Would you prefer that we try to reason out what rules would improve safety here and propose them to the language design team before addressing this issue in the compiler?

**Comment by jaredpar** [comment]
Yeah that seems like the right approach. There were lots of back and forths about how the parent contract got enforced and not sure we can directly apply all of the decisions here. I think there are really two things LDM should consider:

1. Do we want to give parent enforcement here? If so what other attributes do we need to look at here? Basically stop making one off decisions, lets look at them all and make a holistic decision
1. Do we want to mirror the decisions we made for `MemberNotNull` or are there other cases to consider here
